doc/doxygen/headers/automatic_and_symbolic_differentiation.h

```diff
@@ -14,47 +14,389 @@
+ * Three practical examples of auto-differentiation use within a finite-element context
+ * would then be
+ * - the quick prototyping of a new nonlinear formulation without the need to hand-compute
+ *   the linearization itself,
+ * - automatic linearization of finite-element residuals additively formed within complex
+ *   multiphysics frameworks, and
+ * - verification of user-implementations of linearizations for both cell-based calculations
+ *   (e.g. a residual) and those based at a continuum point (e.g. tangents for nonlinear
+ *   constitutive laws).
+ *
+ * There are quite a number of implementations for auto-differentiable numbers. They primarily
+ * fall into two broad categories, namely <em>source code transformation</em> and
+ * <em>operator overloading</em>.
+ * The first method generates new, compilable code based on some input function that, when executed,
+ * returns the derivatives of the input function. The second exploits the capability of <tt>C++</tt>
+ * operator definitions to be overloaded for custom class types. Therefore  a class that represents
+ * such an auto-differentiable number can, following each mathematical operation performed on or
+ * with it, in principle evaluate and keep track of its value as well as that of its directional
+ * derivative(s).
+ * As the libraries exclusively implementing the <em>source code transformation</em> approach
+ * collectively describe highly specialized tools that are to be used as function preprocessors, they
+ * have no direct support within deal.II itself. The latter, however, represent specialized number
+ * types that can be supported through the use of template  metaprogramming in the appropriate context.
+ * Given the examples presented above, this means that the FEValues class (and friends), as well as
+ * the Tensor and SymmetricTensor classes should support calculations performed with these specialized
+ * numbers.
+ * (In theory an entire program could be made differentiable. This could be useful in, for example,
+ * the sentitivity analysis of solutions with respect to input parameters. However, to date this has
+ * not been been tested.)
+ *
+ * Implementations of specialized frameworks based on <em>operator overloading</em> typically fall into
+ * one of three categories. In each, some customized data classes representing the floating point value
+ * of an evaluated function and its derivative(s) by
+ * -# exploiting <em>dual</em> / <em>complex-step</em> / <em>hyper-dual</em> formulations (occasionally
+ *    called <em>tapeless</em> methods),
+ * -# those utilizing <em>taping</em> strategies, and
+ * -# those using compile-time optimization through <em>expression templates</em>.
+ *
+ * To provide some tentative insight into how these various implementations might look like in practice, we
+ * offer the following generic summary of these approaches:
+ * -# The first two <em>tapeless</em> approaches listed above use some variation of a truncated Taylor
```

Comment:
you list three and the second is `taping`, so which of those are meant here?

Reply:
No, this was referring to "dual" and "complex-step" numbers (with the third being "hyper-dual"). I've clarified this in the text now.